linux/src/threadpool_linux.c

```diff
@@ -217,29 +194,25 @@ static int threadpool_work_func(void* param)
                     if (work_function != NULL)
                     {
                         work_function(work_function_ctx);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_039: [ threadpool_work_func shall acquire the shared SRW lock by calling srw_lock_acquire_shared. ]*/
+
                         srw_lock_acquire_shared(threadpool->srw_lock);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_040: [ threadpool_work_func shall save pending_work_item_count_ptr is not NULL in is_pending_work_item_count_ptr_not_null. ]*/
                         bool is_pending_work_item_count_ptr_not_null = (NULL != threadpool->task_array[current_index].pending_work_item_count_ptr);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_041: [ threadpool_work_func shall release the shared SRW lock by calling srw_lock_release_shared. ]*/
                         srw_lock_release_shared(threadpool->srw_lock);
-                        /* Codes_SRS_THREADPOOL_LINUX_05_042: [ If the is_pending_work_item_count_ptr_not_null is TRUE then: ]*/
+
                         if (is_pending_work_item_count_ptr_not_null)
-                        {
-                            /* Codes_SRS_THREADPOOL_LINUX_05_043: [ threadpool_work_func shall acquire the exclusive SRW lock by calling srw_lock_acquire_exclusive. ]*/
+                        {                            
                             srw_lock_acquire_exclusive(threadpool->srw_lock);
```

Comment:
>srw_lock_acquire_exclusive(threadpool->srw_lock); [](http://example.com/codeflow?start=28&length=49)

Why are we taking the lock exclusively here?
This lock was meant to be acquired exclusively only when we resize the task array.
And it is meant to be acquired shared when any operation is performed on the items in the array.

Reply:
We had shared lock and Helgrind was complaining, changing it to exclusive solved the issue since pending_work_item_count_ptr is also initialized in other threads at the index of thread_array for every threadpool_schedule_work_item.